sanic/asgi.py

```diff
@@ -77,35 +76,33 @@ async def shutdown(self) -> None:
         in sequence since the ASGI lifespan protocol only supports a single
         shutdown event.
         """
-        await self.asgi_app.sanic_app._server_event("shutdown", "before")
-        await self.asgi_app.sanic_app._server_event("shutdown", "after")
+        await self.sanic_app._server_event("shutdown", "before")
+        await self.sanic_app._server_event("shutdown", "after")
 
-    async def __call__(
-        self, scope: ASGIScope, receive: ASGIReceive, send: ASGISend
-    ) -> None:
-        message = await receive()
-        if message["type"] == "lifespan.startup":
-            try:
-                await self.startup()
-            except Exception as e:
-                error_logger.exception(e)
-                await send(
-                    {"type": "lifespan.startup.failed", "message": str(e)}
-                )
-            else:
-                await send({"type": "lifespan.startup.complete"})
-
-        message = await receive()
-        if message["type"] == "lifespan.shutdown":
-            try:
-                await self.shutdown()
-            except Exception as e:
-                error_logger.exception(e)
-                await send(
-                    {"type": "lifespan.shutdown.failed", "message": str(e)}
-                )
-            else:
-                await send({"type": "lifespan.shutdown.complete"})
+    async def __call__(self) -> None:
+        while True:
+            message = await self.receive()
+            if message["type"] == "lifespan.startup":
+                try:
+                    await self.startup()
+                except Exception as e:
+                    error_logger.exception(e)
+                    await self.send(
+                        {"type": "lifespan.startup.failed", "message": str(e)}
+                    )
+                else:
+                    await self.send({"type": "lifespan.startup.complete"})
+            elif message["type"] == "lifespan.shutdown":
+                try:
+                    await self.shutdown()
+                except Exception as e:
+                    error_logger.exception(e)
```

Comment:
Doesn't this shows the error message twice?

Reply:
Well, yeah, kinda. It sends the message back to whatever the server logger is using. But this sends to Sanic's internal error logger including the traceback. There is an API to configure Sanic's loggers independent of whatever ASGI server is running, which may be logging to different locations.